qcodes_contrib_drivers/drivers/ERAInstruments/erasynth.py

```diff
@@ -0,0 +1,902 @@
+    "S1": "sweep_start",
+    "S2": "sweep_stop",
+    "S3": "sweep_step",
+    "S4": "sweep_dwell",
+    "P9": "phase_noise_mode",
+    "PEW": "wifi_mode",
+    "PES0": "wifi_sta_ssid",
+    "PEP0": "wifi_sta_password",
+    "PES1": "wifi_ap_ssid",
+    "PEP1": "wifi_ap_password",
+    "PEI": "wifi_ip_address",
+    "PEN": "wifi_subnet_address",
+    "PEG": "wifi_gateway_address",
+}
+"""
+A mapping used to read back certain commands to ensure they have been set.
+This is necessary due to non-deterministic communication times.
+"""
+
+class ERASynthBase(VisaInstrument):
+    r"""
+    A Base class for the ERASynth/ERASynth+/ERASynth++ instruments.
+
+    Example:
+
+    .. code-block::
+
+        from qcodes import Instrument
+        from qcodes_contrib_drivers.drivers.ERAInstruments import ERASynthPlus
+
+        Instrument.close_all()
+
+        # list communication ports
+        ERASynthPlus.print_pyvisa_resources()
+
+        # Instantiate the instrument
+        lo = ERASynthPlus("ERASynthPlus", 'ASRL/dev/cu.usbmodem14101::INSTR')
+
+        lo.off()  # Turn off the output
+
+        # print updated snapshot once to make sure the snapshot will be up-to-date
+        # takes a few seconds
+        lo.print_readable_snapshot(update=True)
+
+        # Configure the local oscillator
+        lo.ref_osc_source("int")  # Use internal reference
+        lo.frequency(4.7e9)
+        lo.power(10)  # Set the amplitude to 10 dBm
+        lo.on()  # Turn on the output
+
+
+    .. seealso::
+
+        The raw serial commands can be found here:
+        https://github.com/erainstruments/erasynth-docs/blob/master/erasynth-command-list.pdf
+
+        ``self.visa_handle`` can be used to interact directly with the serial pyvisa
+        communication.
+    """
+
+    @staticmethod
+    def print_pyvisa_resources() -> None:
+        """Utility to list all."""
+        resource_manager = pyvisa.ResourceManager()
+        print(resource_manager.list_resources())
+
+    def _prep_communication(self):
+        """Makes sure the instrument config is compatible with this driver."""
+        self.debug_messages_en(False)  # Print less messages to improve communication
+        self.wifi_off()  # print less messages to improve communication
+
+    def __init__(self, name: str, address: str, **kwargs):
+        """
+        Create an instance of the instrument.
+
+        Args:
+            name: Instrument name.
+            address: Used to connect to the instrument.
+                Run :meth:`.ERASynthBase.print_pyvisa_resources` to list available list.
+        """
+        super().__init__(name=name, address=address, terminator="\r\n", **kwargs)
+
+        # ##############################################################################
+        # Standard LO parameters
+        # ##############################################################################
+
+        # NB `initial_value` is not used because that would make the initialization slow
```

Comment:
note that initial_cache_value exists, if it can be useful for this parameter

Reply:
Thanks for heads up! It is not used either because the instrument can be in an unknown state so better to have it as `None` in the beginning.

PS Overall the implementation of the serial communication of this instrument (in the physical instrument itself) is a bit messy and complicated to deal with. Unfortunately it was not developed with reliability and automation in mind 🥲